src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -2153,14 +2445,50 @@ new_open_pthread(const char *pathname, int oflags, ...)
 	return rc;
 }
 
+/* Search a fd in fd hash table. Remove it in case it is found. Also free the fake fd.
+ * Return true if fd is found. Return false if fd is not found.
+ */
+static bool
+remove_fd_compatible(int real_fd)
+{
+	struct ht_fd *fd_ht_obj;
+	d_list_t     *rlink;
+
+	rlink = d_hash_rec_find(fd_hash, &real_fd, sizeof(int));
+	if (rlink == NULL)
+		return false;
+
+	fd_ht_obj = fd_obj(rlink);
+	/* close fake fd */
+	if (fd_ht_obj->fake_fd >= FD_DIR_BASE)
+		free_dirfd(fd_ht_obj->fake_fd - FD_DIR_BASE);
+	else
+		free_fd(fd_ht_obj->fake_fd - FD_FILE_BASE, false);
+	/* remove fd from hash table */
+	d_hash_rec_decref(fd_hash, rlink);
+	return true;
+}
+
 static int
 new_close_common(int (*next_close)(int fd), int fd)
 {
-	int fd_directed;
+	int fd_directed, rc;
 
 	if (!hook_enabled)
 		return next_close(fd);
 
+	if (compatible_mode && fd < FD_FILE_BASE) {
+		remove_fd_compatible(fd);
+		if (fd < DAOS_MIN_FD && daos_inited) {
+			rc = dup2(DAOS_DUMMY_FD, fd);
```

Comment:
This needs updating.
```suggestion
			rc = dup2(fd_dummy, fd);
```

Reply:
Thank you for catching this! It is fixed now.